torch/onnx/_internal/fx/passes/type_promotion.py

```diff
@@ -933,3 +968,438 @@ def get_rule(
+    ), "Expected at least 1 traced op, got 0"
+
+    new_op_overload = op_trace_dispatch_mode.traced_ops[0]
+    assert isinstance(
+        new_op_overload, torch._ops.OpOverload
+    ), f"Expected OpOverload, got {type(new_op_overload)}"
+    assert (
+        new_op_overload.overloadpacket == op
+    ), f"Expected same OpOverload packet, got {new_op_overload.overloadpacket} != {op}"
+
+    return new_op_overload
+
+
+class _TypePromotionInterpreter(torch.fx.Interpreter):
+    """Interpreter that inserts type promotion for each node."""
+
+    def __init__(
+        self,
+        diagnostic_context: diagnostics.DiagnosticContext,
+        module: torch.fx.GraphModule,
+        type_promotion_table: TypePromotionTable,
+    ):
+        super().__init__(module)
+        self.diagnostic_context = diagnostic_context
+        self.type_promotion_table = type_promotion_table
+
+    def _run_node_and_update_meta_val(self, node) -> Any:
+        """Run a node and update node.meta["val"] with the output value.
+
+        This should be used on new nodes or nodes that have been modified.
+        By default `Interpreter.run_node` does not update `node.meta["val"]`.
+        """
+        out = super().run_node(node)
+        self.env[node] = out
+        node.meta["val"] = proxy_tensor.extract_val(out)
+        return out
+
+    @_beartype.beartype
+    def _create_node(
+        self,
+        graph: torch.fx.Graph,
+        op_type: str,
+        target: torch.fx.node.Target,
+        args: tuple,
+        kwargs: dict,
+    ) -> torch.fx.Node:
+        """Create a node and set its metadata."""
+        assert op_type in (
+            "call_function",
+            "call_method",
+            "get_attr",
+            "call_module",
+            "placeholder",
+            "output",
+        ), f"Unexpected op_type: {op_type}"
+        node = getattr(graph, op_type)(target, args, kwargs)
+        self._run_node_and_update_meta_val(node)
+        node.meta.update(
+            (k, v)
+            for k, v in fx_traceback.get_current_meta().items()
+            if k not in node.meta
+        )
+        return node
+
+    @_beartype.beartype
+    def _rerun_node_after_type_promotion(
+        self,
+        diagnostic: diagnostics.Diagnostic,
+        node: torch.fx.Node,
+        expected_out_dtype: torch.dtype,
+    ) -> None:
+        """Rerun a node after type promotion and update node.meta["val"] with the output value."""
+        node_val = node.meta.get("val", None)
+        assert node_val is not None, f"Node {node} node.meta['val'] is not set."
+        args, kwargs = self.fetch_args_kwargs_from_env(node)
+        target = node.target
+        assert isinstance(
+            target, torch._ops.OpOverload
+        ), f"Expected OpOverload, got {type(target)}"
+        node.target = find_compatible_op_overload(target.overloadpacket, args, kwargs)
+
+        new_node_val = self._run_node_and_update_meta_val(node)
```

Comment:
No because this is not a new node. Frankly this is case by case. In this pass for modified nodes we only need to update 'val' in 'meta'. But if you prefer I can simplify the api to always copy all other meta info.

Largely this is the problem with freestyle design of `fx.Node` meta. Devs need to be super aware of its content to correctly maintain it. And meta is very important as it serves as source info for numerous high level apis.

Reply:
I will isolate the change in a follow up PR.

Reply:
https://github.com/pytorch/pytorch/pull/104720